Faster Array>>asX and Array>>as:

privat opened the pull request:
Important low-level change where conversion from Array are made faster. (up to x3 (at least) for `asOrderedCollection`)

This is useful as `{1. 2. 3} asFoo` (or `{1. 2. 3} as: Foo`) is the current de facto syntax to have initialized collections of various types with user defined content.

Note that subclasses of Collection that redefine `newFrom:` should also redefine `newFromArray:` or bad things could happen.
This is because the default (inherited) implementations of `newFromArray:` uses `new:` and `add:` (or `at:put:` for ArrayedCollection).

privat committed:
Collection: add newFromArray facility

privat committed:
Array: activate fast asX: conversion

Will this break something?

astares commented on src/Collections-Abstract/Collection.class.st:
```diff
@@ -22,6 +22,37 @@ Collection class >> isAbstract [
+	"Fast initialization with the items of a given array.
+	This initializes elements faster that the generic withAll: or newFrom: methods.
+
+	The main selling point is that dyamic arrays like {1. 2. 3} are really fast in Pharo.
```
typo: dynamic

privat committed:
fix typo in comment of newFromArray:

privat committed:
Forgot to commit Object>>asNewArray

privat committed:
Array>>as: forgot to return self if already an Array

Did that break only 4 tests?

jordanmontt commented:
> Note that subclasses of Collection that redefine newFrom: should also redefine newFromArray: or bad things could happen.
This is because the default (inherited) implementations of newFromArray: uses new: and add: (or at:put: for ArrayedCollection).

Maybe you can add this in a method comment like that the developers can know :)

privat committed:
Redefine RunArray>>newFromArray:

note: the inherited implementation does not work because `RunArray class>>new:` initializse a collection of a given size filled with nil. This behavior feels highly dubious but fixing this is out of scope for the commit (or the PR)

privat committed:
Collection class>>newFromArray: document when and how redefine it in subblasses

privat committed:
RunArray>>NnewFrom Fix executable comment

hogoww commented:
Presentation talking about this topic (thanks Jean!)
https://rmod-files.lille.inria.fr/Team/PharoPresentations/2022-DynamicInitializationOfCollection-Privat.pdf

hogoww reviewed (changes requested):
nevermind, I misread :D

Ducasse commented on src/Collections-Abstract/Collection.class.st:
```diff
@@ -296,6 +330,18 @@ Collection >> asIdentitySet [
+{ #category : #converting }
+Collection >> asNewArray [
+
+	"Like asArray: but return a copy is self is already an Array.
```
Like asArray: but return a copy if self is already an Array.

Ducasse commented on src/Kernel/Object.class.st:
```diff
@@ -140,6 +140,17 @@ Object class >> newFrom: aSimilarObject [
 ]
 
+{ #category : #'instance creation' }
+Object class >> newFromArray: anArray [
```
Do we want this on Object?
I would not.

privat replied:
As the comment of the method says, this is only for backward compatibility.

Some existing clients can (directly or not) rely on the fact that `{1. 2. 3} as: SomeClass` calls back `SomeClass>>newFrom:`. With the change proposed in the PR the likely DNU selector `newFromArray:` will be called instead. So to have a semantically equivalent change, I just globally forward `newFromArray:` to `newFrom:` unless a better `newFromArray:` is redefined.

Side note: I'm not sure about the *backward compatibility policy* of Pharo (is there an implicit or an explicit one?) I can adapt the PR to what you prefer: I can drop `Object class>>newFromArray:` (and zut for guys that are on the way). I can also harden the current proposal for compatibility: move specific behavior Collection>>newFromArray into selected subclasses (so weird subclasses of Collection that redefine newFrom: or withAll: will not break). I can even add type guards to that weird subclasses of OrderedCollection or Array do not feel the change (fragile base class bla bla bla).

Other note: the rationale and laws of Object-level methods like `Object>>as:` or `Object class>>newFrom:` are not completely evident for me. While this is not the point of the PR to discuss these, feel free to say if I broke some beahviors or rules.

Ducasse commented on src/Collections-Sequenceable/Array.class.st:
```diff
@@ -136,13 +136,48 @@ Array class >> new: sizeRequested [
 
+{ #category : #converting }
+Array >> asDictionary [
+
```
I would not add asDictionary because it is really strange to me and can only work if we have binding in the array.

guillep replied:
I'm not against this change per se.
@Ducasse , asDictionary already exists. And works only for collections with associations.
So, this change is semantically equivalent.

Of course, the rationale behind this is speed, but if the change is questioned for Dictionaries, then all the rest of the changes are too. The question is: Do we want `{ a->b. c->d } asDictionary` to be preferred to  `Dictionary new at: a put: b; at: c put: d; yourself`?.

Regarding performance, however, this only cuts one message send, right? Instead of

```
anArray asDictionary
anArray as: Dictionary
Dictionary newFromArray: anArray
...
```

It would be:
```
anArray asDictionary
Dictionary newFromArray: anArray
...
```

right?
Is there much to gain in here?

privat replied:
Since there was an existing `asDictionary` in Array, I just adapted it to use the fast path from `Collection class>>newFromArray:` and be equivalent with `as: Dictionnary` (as `Array>>as:` uses the fast path by default).

Note: see the doc of `Collection class>>newFromArray:` for where is the gain

privat replied:
I forgot the rule of  ``do not comment before your first coffee''.

It seems that there was in fact no previous definition of `asDictionary` therefore I'm not sure why I added it.
I will remove it since there is almost no gain as noted @guillep

MarcusDenker replied:
There is #asDictionary in Pharo (on Collection). It has 90 senders.

MarcusDenker replied:
I propose to make *this* PR solely about speeding up what we have and do change as part of another step.

privat replied:
I removed asDictionary

privat committed:
typo in the comment of asNewArray

privat committed:
remove unneeded Array>>asDictionary

MarcusDenker commented:
Failing tests not related:

- [CI] Failing test: ProperMethodCategorizationTest>>testNoUncategorizedMethods #11669
